e220/e220.go

```diff
@@ -0,0 +1,349 @@
+		d.m1.High()
+		d.m0.Low()
+	case Mode3:
+		d.m1.High()
+		d.m0.High()
+	default:
+		return fmt.Errorf("Invalid mode: %d", mode)
+	}
+
+	return nil
+}
+
+const (
+	writeCmd = 0xC0
+	readCmd  = 0xC1
+)
+
+// ReadRegister reads and returns length register values from startAddr
+func (d *Device) ReadRegister(startAddr uint8, length uint8) ([]byte, error) {
+	response, err := d.writeRegister(readCmd, startAddr, length, []byte{})
+	if err != nil {
+		return nil, err
+	}
+	// response must start at 0xC1 and be the length requested
+	if (len(response) != int(length+3)) || (response[0] != 0xC1) {
+		return nil, fmt.Errorf("unexpected response: %X", response)
+	}
+
+	return response[3:], nil
+}
+
+// WriteRegister writes the register value from startAddr to the value of params
+func (d *Device) WriteRegister(startAddr uint8, params []byte) error {
+	response, err := d.writeRegister(writeCmd, startAddr, uint8(len(params)), params)
+	if err != nil {
+		return err
+	}
+	// response must start with C1 and the subsequent values must be the same as the requested values
+	if (response[0] != 0xC1) || ((response[1]) != startAddr) ||
+		(int(response[2]) != len(params)) || !bytes.Equal(response[3:], params) {
+		return fmt.Errorf("unexpected response: want=%X%X got=%X",
+			[]byte{0xC1, startAddr, byte(len(params))}, params, response)
+	}
+
+	return nil
+}
+
+func (d *Device) writeRegister(cmd, startAddr, length uint8, params []byte) ([]byte, error) {
+	d.payload[0] = cmd
+	d.payload[1] = startAddr
+	d.payload[2] = length
+	if len(params) > writingRegisterSize {
+		return nil, fmt.Errorf("params must be greater than or equal to %d: got=%d", writingRegisterSize, len(params))
+	}
+	copy(d.payload[3:], params)
+	writeLength := uint8(3)
+	if cmd == writeCmd {
+		writeLength += length
+	}
+	// Wait for device is ready
+	ctxWrite, cancelWrite := context.WithTimeout(context.Background(), 5000*time.Millisecond)
+	defer cancelWrite()
+	for !d.IsReady() {
+		select {
+		case <-ctxWrite.Done():
+			return nil, fmt.Errorf("writing register timed out")
+		default:
+		}
+		runtime.Gosched()
+	}
+	_, err := d.uart.Write(d.payload[:writeLength])
+	if err != nil {
+		return nil, err
+	}
+
+	ctxRead, cancelRead := context.WithTimeout(context.Background(), 5000*time.Millisecond)
+	defer cancelRead()
+	readIndex := 0
+	for {
+		if d.uart.Buffered() > 0 {
+			d.payload[readIndex], err = d.uart.ReadByte()
+			if err != nil {
+				return nil, err
+			}
+			readIndex++
+			if readIndex == 3+int(length) {
+				break
+			}
+		}
+		select {
+		case <-ctxRead.Done():
+			return nil, fmt.Errorf("waiting for response from E220 timed out")
+		default:
+		}
+		runtime.Gosched()
+	}
+	// Wait for a while because the next register access
+	// cannot be received immediately after register access
+	time.Sleep(50 * time.Millisecond)
+
+	return d.payload[:readIndex], nil
+}
+
+// WriteConfig writes configuration values to E220
+func (d *Device) WriteConfig(cfg Config) error {
+
+	cfg.Validate()
```

Comment:
Validate should return an `error` type itself. check out the `errors.Join(errSlice...)` function which does pretty much what you are doing below.